compiler/transforms/acc_dedup.py

```diff
@@ -56,6 +52,123 @@ def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+                return
+            prev_op = prev_op.prev_op
+        if prev_op is None:
+            return
+
+        state = dict(prev_op.iter_params())
+        state.update(dict(op.iter_params()))
+
+        rewriter.replace_op(
+            prev_op,
+            new_setup := acc.SetupOp(
+                state.values(), state.keys(), op.accelerator, prev_op.in_state
+            ),
+        )
+        op.out_state.replace_by(new_setup.out_state)
+
+        rewriter.erase_matched_op()
+
+
+class ElideEmptySetupOps(RewritePattern):
+    """
+    remove setup ops that set zero parameters, but only if they have both a in- and an out-state
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+        if len(op.values) == 0 and op.in_state is not None and op.out_state is not None:
+            op.out_state.replace_by(op.in_state)
+            rewriter.erase_matched_op()
+
+
+class PullSetupOpsOutOfLoops(RewritePattern):
+    """
+    Tries to pull setup operations out of loop nests by:
+    1. Inspecting the source block of all operands
+    2. Get all operands that originate not from the block the current operation is in or its descendants (X)
+    3. Clone the operation, remove all ops that are not in X
+    4. Insert the cloned op right before the loop
+    5. Remove all operands from the original op that are in X
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+        # don't apply to setups not inside for loops
+        loop_op = op.parent_op()
+        if not isinstance(loop_op, scf.For):
+            return
+
+        # only do this for the first setup op in the loop
+        if op.in_state is None or op.in_state.owner != loop_op.body.block:
+            return
+
+        # iterate over all setups inside this loop and check if their values are loop-invariant or not
+        save_values: set[str] = set()  # loop invariant values
+        unsafe_vals: set[str] = set()  # loop dependent values
+        state: dict[str, SSAValue] = {}
+        for setup in all_setup_states_in_region(
+            loop_op.body, op.accelerator.data, recurse=True
+        ):
+            for key, val in setup.items():
+                # a value is "unsafe" if it originates inside the loop even on one setup call
+                if val_is_defined_in_block(val, loop_op.body.block):
+                    unsafe_vals.add(key)
+                # and also if it changes at any point in the loop
+                elif key in state and state[key] != val:
+                    unsafe_vals.add(key)
+                else:
+                    state[key] = val
+                    save_values.add(key)
+        # remove all unsafe vals form potentially safe values
+        # also pick a deterministic, fixed order for the rest of the rewrite
+        loop_invariant_options = tuple(sorted(save_values - unsafe_vals))
```

Comment:
If a value is first set to a loop-invariant value, and later set to an unsafe value, it will first be added to `safe_vals` and later to `unsafe_vals`. Therefore it is necessary to subtract the unsafe vals from the safe ones after the loop.